### PR TITLE
Add Sonic's base-fee computation algorithm

### PR DESCRIPTION
**Pull request by HerbertJordan**
This PR introduces Sonic's base-fee adjustment algorithm. The algorithm is a generalization of the fee market price algorithm described in [EIP-1559](https://ethereum.github.io/abm1559/notebooks/eip1559.html) supporting a variable block-time period.

The maximum increase and decrease is limited to approximately ±12.5% per 15 seconds, analog to Ethereum's implementation.

### CODE REVIEW

File: gossip/gasprice/base_fee.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	//   - targetRate is the target gas rate per second at which prices are stable
+	//   - duration is the time in seconds between the parent and grand-parent blocks
+	//
+	// All computations are carried out using integers to avoid floating point errors.
+	// To that end, terms are re-arranged to fit the following shape:
+	//
+	//               newPrice := oldPrice * e^(numerator/denominator)
+	//
+	// where numerator and denominator are integers. The final value is then computed
+	// using an approximation of the this function based on a Taylor expansion around 0.
```
Comment (kjezek) [inline gossip/gasprice/base_fee.go:38]:
```suggestion
	// using an approximation of this function based on a Taylor expansion around 0.
```

**Review by simonlechner — APPROVED** [review]
Looks good, the comments are definitely helpful.

File: inter/time.go [hunk 2]
```diff
@@ -9,6 +9,8 @@ import (
 type (
 	// Timestamp is a UNIX nanoseconds timestamp
 	Timestamp uint64
+	// Duration is a UNIX nanoseconds duration
+	Duration uint64
```
Comment (LuisPH3) [inline inter/time.go:13]: I am not quite sure about this declaration.
- The standard library offers time. Duration, which is an int64, but end up casting this duration to int64 to construct `big.Int` quite often. 
- Additionally, this file seems to declare a specific timestamp in unix format, not a library of time utils.

File: gossip/gasprice/base_fee.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+	oldPrice := new(big.Int).Set(parent.BaseFee)
+
+	// If the time gap between the parent and this block is zero or more
+	// than 60 seconds, something significantly disturbed the chain and we
+	// keep the BaseFee constant.
+	duration := uint64(parent.Duration)
```
Comment (LuisPH3) [inline gossip/gasprice/base_fee.go:45]: nit: redundant cast

**Review by jenikd — APPROVED** [review]
Thanks, looks good

**Review by LuisPH3 — APPROVED** [review]
Lgtm, just small nitpick about new introduced type.

**Review by LuisPH3 — APPROVED** [review]

**Review by jenikd — APPROVED** [review]

**Pull request merged**